lib/AST/AutoDiff.cpp

```diff
@@ -308,21 +330,48 @@ GenericSignature autodiff::getConstrainedDerivativeGenericSignature(
   auto &ctx = originalFnTy->getASTContext();
   auto *diffableProto = ctx.getProtocol(KnownProtocolKind::Differentiable);
   SmallVector<Requirement, 4> requirements;
-  for (unsigned paramIdx : diffParamIndices->getIndices()) {
-    // Require differentiability parameters to conform to `Differentiable`.
-    auto paramType = originalFnTy->getParameters()[paramIdx].getInterfaceType();
-    Requirement req(RequirementKind::Conformance, paramType,
+
+  auto addRequirement = [&](CanType type) {
+    Requirement req(RequirementKind::Conformance, type,
                     diffableProto->getDeclaredInterfaceType());
     requirements.push_back(req);
     if (isTranspose) {
       // Require linearity parameters to additionally satisfy
       // `Self == Self.TangentVector`.
-      auto tanSpace = paramType->getAutoDiffTangentSpace(lookupConformance);
-      auto paramTanType = tanSpace->getCanonicalType();
-      Requirement req(RequirementKind::SameType, paramType, paramTanType);
+      auto tanSpace = type->getAutoDiffTangentSpace(lookupConformance);
+      auto tanType = tanSpace->getCanonicalType();
+      Requirement req(RequirementKind::SameType, type, tanType);
       requirements.push_back(req);
     }
+  };
+
+  // Require differentiability parameters to conform to `Differentiable`.
+  for (unsigned paramIdx : diffParamIndices->getIndices()) {
+    auto paramType = originalFnTy->getParameters()[paramIdx].getInterfaceType();
+    addRequirement(paramType);
+  }
+
+  // Require differentiability results to conform to `Differentiable`.
+  SmallVector<SILResultInfo, 2> originalResults;
+  getSemanticResults(originalFnTy, diffParamIndices, originalResults);
+  for (unsigned resultIdx : diffResultIndices->getIndices()) {
+    // Handle formal original result.
+    if (resultIdx < originalFnTy->getNumResults()) {
+      auto resultType = originalResults[resultIdx].getInterfaceType();
+      addRequirement(resultType);
+      continue;
+    }
+    // Handle original semantic result parameters.
+    // FIXME: Constraint generic yields when we will start supporting them
+    auto resultParamIndex = resultIdx - originalFnTy->getNumResults();
```

Comment:
Technically, checking semantic result parameters and adding corresponding requirements here is redundant as those are always wrt params. However, I added for the sake of completeness. And we will add here yields in a similar way when they will be supported